src/aof.c

```diff
@@ -928,7 +1337,114 @@ int loadAppendOnlyFile(char *filename) {
+    serverAssert(am != NULL);
+    int ret = C_OK;
+    long long start;
+    off_t base_aof_size = 0, incr_aofs_size = 0, size;
+    sds aof_name;
+
+    /* If there is no information about BASE AOF and INCR AOF in aofManifest, then 
+     * there is only one possibility: the aof manifest file does not exist, we may 
+     * be starting from an old redis version. so we must fall back to the previous 
+     * loading mode.
+     */
+    if (am->base_aof_info == NULL && listLength(am->incr_aof_list) == 0) {
+        struct redis_stat sb;
+        /* If the `server.aof_filename` file does not exist, we will directly return 
+         * AOF_NOT_EXIST, and redis will ignore this error. */
+        if (redis_stat(server.aof_filename, &sb) != 0) {
+            return AOF_NOT_EXIST;
+        }
+
+        /* If the server.aof_filename file exists, we manually construct a BASE 
+         * type aofInfo and add it to aofManifest. In this way, we can reuse the 
+         * following code to load this AOF file. */
+        aofInfo *ai = aofInfoCreate();
+        ai->file_name = sdsnew(server.aof_filename);
+        ai->file_seq = 1;
+        ai->file_type = AOF_FILE_TYPE_BASE;
+        am->base_aof_info = ai;
+        am->curr_base_aof_seq = 1;
+        am->dirty = 1;
+
+        int ret = persistAofManifest(am);
+        if (ret == C_ERR) {
+            return AOF_FAILED;
+        }
+    }
+
+    serverAssert(am->base_aof_info != NULL ||
+        listLength(am->incr_aof_list) != 0);
+
+    /* Load BASE AOF if needed. */
+    if (am->base_aof_info) {
+        serverAssert(am->base_aof_info->file_type == AOF_FILE_TYPE_BASE);
+        aof_name = (char*)am->base_aof_info->file_name;
+        size = getAppendOnlyFileSize(aof_name);
+        start = ustime();
+        /* Note: Although we use `startLoadingRdbFile` here, we may actually 
+         * be loading an AOF file, such as when we upgrade from an old version redis. */
+        startLoadingRdbFile(size, aof_name, RDBFLAGS_AOF_PREAMBLE);
+        ret = loadSingleAppendOnlyFile(aof_name);
+        stopLoading(ret == AOF_OK);
+
+        if (ret == AOF_OK) {
+            base_aof_size = size;
+            serverLog(LL_NOTICE, "DB loaded from append only file %s: %.3f seconds",
+                aof_name, (float)(ustime()-start)/1000000);
+        }
+
+        /* If an AOF exists in the manifest but not on the disk, we consider 
+         * this to be a fatal error. */
+        if (ret == AOF_NOT_EXIST) ret = AOF_FAILED;
+
+        if (ret != AOF_OK && ret != AOF_EMPTY) {
+            return ret;
+        }
+    }
+
+    /* Load INCR AOFs if needed. */
+    if (listLength(am->incr_aof_list) > 0) {
+        listNode *ln;
+        listIter li;
+
+        /* Here we calculate the total size of all INCR AOF files in advance, it 
+         * will be set to `server.loading_total_bytes`. */
+        incr_aofs_size = getIncrAppendOnlyFilesSize(server.aof_manifest);
+        startLoadingIncrAofFiles(incr_aofs_size);
```

Comment:
@oranagra I added a `startLoadingIncrAofFiles` function to set the size of the entire INCR AOF, and then in the `loadSingleAppendOnlyFile` function, `loadingIncrAofProgress` will be used to update the progress of the loading. Unlike `loadingRdbProgress` (originally `loadingProgress`), `loadingIncrAofProgress` receives a relative size (not a Absolute file offset).

Reply:
I meant just one startLoading for the whole thing (not one for the base and one for all incr files)

regarding the new functions that's ok, just that maybe the function naming should be neutral and just state the fact of what they do (relative or absolute instead of RDB or AOF)

Reply:
Modified

Reply:
@oranagra  Please review this part again，thanks.